app/views/pages/carte_inscrits.erb

```diff
@@ -0,0 +1,53 @@
+<%= content_for(:append_to_head) do %>
+<meta content="index;follow" name="robots"/>
+<% end %>
+
+<style>
+  #map-container {
+    height: 500px;
+    position: relative; /* used to set a reference for the children */
+  }
+
+  #map, #deck-canvas {
+    /* children will take the full space of the container */
+    position: absolute;
+    top: 0;
+    left: 0;
+    width: 100%;
+    height: 100%;
+  }
+</style>
+
+<h1><%= VaccinationCenter.count %> centres sur CovidListe, et des volontaires dans tous les d&eacute;partements</h1>
+
+<div id="map-container">
+  <div id="map"></div>
+  <canvas id="deck-canvas"></canvas>
+</div>
+
+<script language="javascript">
+  window.addEventListener('load', () => {
+    const data = <%=
+      vaccination_centers = VaccinationCenter.all.map do |v|
+        {
+          'coordinates' => [v.lon, v.lat],
+          'nom' => v.name,
+        }
+      end
+      users_by_dept = Rails.cache.fetch("users_by_dept", expires_in: 1.hours) do
+        User.all
+          .reject { |u| u.departement_code.nil? }
+          .group_by { |u| u.departement_code }
+          .transform_values(&:count)
+          .reject { |_, count| count < 10 }
+      end
+      objects = {
+        'vaccinationCenters' => vaccination_centers,
+        'usersByDept' => users_by_dept
+      }
+      raw(objects.to_json)
```

Comment:
Les données sont passées du monde rails au monde JS ici

Reply:
Effectivement, ici la requête SQL est "select * from users" et ensuite tu filtres et transforme le résultat in-memory en ruby. Ca risque d'être assez lourd. Peut être qu'il vaut mieux faire directement l'ensemble le group by en SQL avec une seule requête

Reply:
```
select 
substring(geo_citycode, 1, 2) as dept,
count(*) as users_count
from users 
where 
confirmed_at is not null
group by 1
having count(*) > 10
order by 2 desc
```

Reply:
y a le cas de la corse à gérer

Reply:
```
select 
case 
when substring(geo_citycode, 1, 2) in ('2A', '2B') then substring(geo_citycode, 1, 2)
when substring(geo_citycode, 1, 2)::int <= 95  
then substring(geo_citycode, 1, 2)
else substring(geo_citycode, 1, 3) end as dept,
count(*) as users_count
from users 
where 
confirmed_at is not null
and geo_citycode is not null
group by 1
having count(*) > 10
order by 2 desc
```

Reply:
et pour lancer la requete raw SQL en ruby tu peux faire:

```
sql = <<~SQL.tr("\n", ' ').squish
    select 
    case when substring(geo_citycode, 1, 2) in ('2A', '2B') then substring(geo_citycode, 1, 
    2) 
    when substring(geo_citycode, 1, 2)::int <= 95 then substring(geo_citycode, 1, 2) 
    else substring(geo_citycode, 1, 3) end as dept, 
    count(*) as users_count 
    from users 
     where confirmed_at is not null 
     and geo_citycode is not null 
     group by 1 having count(*) > 10 
     order by 2 desc
SQL
query = ActiveRecord::Base.send(:sanitize_sql_array, [sql])
results = ActiveRecord::Base.connection.execute(query).to_a
```

Reply:
Si il y a autant de Ruby, ce serait mieux de la mettre soit dans le controlleur (pour les variables d'instances pertinentes) et/ou des helpers 🙏

Reply:
Top je vais faire ca (cc @EmCousin car cela rejoint ton commentaire)

Reply:
A tester et voir si la méthode plain ruby + cache fonctionne. J'ai testé la requête SQL sur la base de production, c'est presque instantané

Reply:
J'ai essayé de faire:
```ruby
ActiveRecord::Base.connection.execute(sql)
```
mais j'obtiens l'erreur `no _dump_data is defined for class PG::Result`. Des idées?

Reply:
Je viens de faire la modif où tout est en SQL.
J'ai aussi enlevé le code qui extrait le departement du controlleur ruby pour eviter les doublons.